teal/packer/plugin.tl

```diff
@@ -0,0 +1,238 @@
+  end
+
+  if user.rev then
+    name = name .. '@' .. user.rev
+  end
+
+  return name
+end
+
+local function remove_ending_git_url(url: string): string
+  return vim.endswith(url, '.git') and url:sub(1, -5) or url
+end
+
+local function normspec(x: string|M.UserSpec): M.UserSpec
+  return x is string and { x } or x as M.UserSpec
+end
+
+local function normcond(x: string|{string}): {string}
+  if x is string then
+    return { x }
+  end
+  return x as {string}
+end
+
+local function normkeys(x: string|{{string,string}}): {{string,string}}
+  if x is string then
+    return {{ '', x as string }}
+  end
+  return x as {{string,string}}
+end
+
+local function normrun(x: string|function|{string|function}): {string|function}
+  if x is function or x is string then
+    return { x }
+  end
+  return x as {string|function}
+end
+
+--- The main logic for adding a plugin (and any dependencies) to the managed set
+-- Can be invoked with (1) a single plugin spec as a string, (2) a single plugin spec table, or (3)
+-- a list of plugin specs
+local function process_spec(
+  spec0: string|M.UserSpec,
+  plugins: {string:M.Plugin},
+  required_by: M.Plugin
+): {string:M.Plugin}
+  local spec = normspec(spec0)
+
+  if #spec > 1 then
+    local r: {string:M.Plugin} = {}
+    for _, s in ipairs(spec) do
+      r = vim.tbl_extend('error', r, process_spec(s, plugins, required_by))
+    end
+    return r
+  end
+
+  local id = spec[1] as string
+  spec[1] = nil
+
+  if id == nil then
+    log.warn('No plugin name provided!')
+    return {}
+  end
+
+  local name, path = get_plugin_name(id)
+
+  if name == '' then
+    log.warn(fmt('"%s" is an invalid plugin name!', id))
+    return {}
+  end
+
+  if plugins[name] then
+    if required_by then
+      plugins[name].required_by = plugins[name].required_by or {}
+      table.insert(plugins[name].required_by, required_by.name)
+    else
+      log.warn(fmt('Plugin "%s" is specified more than once!', name))
+    end
+
+    return {[name] = plugins[name]}
+  end
+
+  local url, ptype = guess_plugin_type(path)
+
+  local plugin: M.Plugin = {
+    name        = name,
+    full_name   = get_plugin_full_name(name, spec),
+    branch      = spec.branch,
+    rev         = spec.rev,
+    tag         = spec.tag,
+    commit      = spec.commit,
+    keys        = normkeys(spec.keys),
+    event       = normcond(spec.event),
+    ft          = normcond(spec.ft   ),
+    cmd         = normcond(spec.cmd  ),
+    cond        = spec.cond,
+    run         = normrun(spec.run),
+    lock        = spec.lock,
+    url         = remove_ending_git_url(url),
+    type        = ptype,
+    config      = spec.config,
+    required_by = required_by and {required_by.name} or nil,
+    revs        = {}
+  }
+
+  plugins[name] = plugin
+
+  if plugin.opt == nil then
+    plugin.opt = plugin.keys ~= nil
+      or plugin.ft ~= nil
+      or plugin.cmd ~= nil
+      or plugin.event ~= nil
+      or plugin.cond ~= nil
```

Comment:
If the keys change, then the Teal compiler will tell us. Pulling in the conds from `Handlers` will require us to use  `as` and reduce the level of type checking.

The issue is that `Plugin` contains fields names with all the enums in `Handlers.Types` so there's some duplication at the typing level. If we want to avoid that then we would need to nest the conditionals in `Plugin` with something like `Plugin.handlers: {Handlers.Type: any}`, but since the value is different for each it introduces worse problems.

Reply:
A case I'm thinking of here stems from user extensions to the set of handlers. I think allowing this in a structured way could be a good way for `packer` to only need to ship a simple core while allowing users to have all the other features they want to use, without requiring them to be part of `packer` itself. In that case, we do need some way to iterate over the handlers. Maybe designing with extensibility in mind doesn't need to be part of this PR, but I think we will eventually come up against this issue.

I admit, this goal doesn't play nicely with type-level key checking. I'll need to think more about the right way to do this.

Reply:
I think we need to treat the keys in `Plugin` and the handlers as separate things. Currently, a handler is just `function(Plugin, loader)` meaning a handler has access to the entire plugin object, and not just the keys it uses. This allows handlers that aren't strictly defined by a single key and may use several, or none.

Also the typing of the fields of `Plugin` is just for our benefit internally. This is a general trend in Teal where from time-to-time you need to duplicate type defs to get the strongest checks. This kind of duplication never leads to buggy code like it would in Lua, and in fact does the opposite by making things more robust since the compiler has more information.